src/dev/amdgpu/sdma_engine.cc

```diff
@@ -510,9 +510,12 @@ SDMAEngine::decodeHeader(SDMAQueue *q, uint32_t header)
         dmaReadVirt(q->rptr(), sizeof(sdmaAtomic), cb, dmaBuffer);
         } break;
       case SDMA_OP_CONST_FILL: {
-        q->incRptr(sizeof(sdmaConstFill));
-        warn("SDMA_OP_CONST_FILL not implemented");
-        decodeNext(q);
+        DPRINTF(SDMAEngine, "SDMA Constant fill packet\n");
+        dmaBuffer = new sdmaConstFill();
+        cb = new DmaVirtCallback<uint64_t>(
+            [ = ] (const uint64_t &)
```

Comment:
I'm confused by this... I've never seen this before. Can you explain it?

Reply:
This is creating a callback for dmaReadVirt using lambda function (which begins at the [=] part).  The callback is called when the dmaReadVirt call finishes doing all translation and all of the DMA requests have returned from memory (since the size could be arbitrarily large this may result in multiple page-size DMAs we need to wait for).  It will call the constFill function in this class when it calls back